File: Cpp_chunks/stage_1.cpp
class SudokuGrid {
    /*...*/
public:
    void stageOne() {
        int row, col;
        SudokuCell *cell;
        bool affectedCells = false;
        for (row = 0; row < n; row++) {
            for (col = 0; col < n; col++) {
                cell = &Grid[row][col];
                if (!cell->hasValue() && cell->avSet->size == 1) {
                    affectedCells = true;
                    updateCell(row, col, cell->avSet->values()[0]);
                }

            }
        }
        if (affectedCells) {
            stageOne();
        }
    }
};

File: Python/chunks/chapter_2/stage_1/stage_1.py
class SudokuGrid:
    """..."""
    def stage_one(self):
        for row in self.I:
            for col in self.I:
                cell = self.grid[row][col]
                if not cell.has_value() and len(cell.av_set) == 1:
                    for num in cell.av_set:
                        self.update_cell(row, col, num)